ssz/sedes/list.py

```diff
@@ -0,0 +1,119 @@
+    """
+    LENGTH_BYTES = 4
+
+    def __init__(self, element_sedes=None, empty=False):
+        if element_sedes and empty:
+            raise ValueError(
+                "Either one of Element Sedes or Empty has to be specified"
+            )
+
+        elif not element_sedes and not empty:
+            raise ValueError(
+                "Either Element Sedes or Empty has to be specified"
+            )
+
+        # This sedes object corresponds to each element of the iterable
+        self.element_sedes = element_sedes
+        # This empty bool indicates whether this sedes is meant for empty lists
+        self.empty = empty
+
+    def serialize(self, val):
+        if (
+                not isinstance(val, Iterable) or
+                isinstance(val, bytes) or
+                isinstance(val, bytes) or
+                isinstance(val, str)
+        ):
+            raise SerializationError(
+                'Can only serialize Iterable objects, except Dictionaries',
+                val
+            )
+
+        if self.empty and len(val) != 0:
+            raise ValueError(
+                "Empty List Sedes cannot serialize non-empty Iterables"
+            )
+
+        serialized_iterable_string = b"".join(
+            self.element_sedes.serialize(element) for element in val
+        )
+
+        if len(serialized_iterable_string) >= 2 ** (self.LENGTH_BYTES * 8):
+            raise SerializationError(
+                'List too long to fit into {} bytes after serialization'.format(self.LENGTH_BYTES),
+                val
+            )
+        serialized_len = len(serialized_iterable_string).to_bytes(self.LENGTH_BYTES, 'big')
+
+        return serialized_len + serialized_iterable_string
+
+    def deserialize_segment(self, data, start_index):
+        """
+        Deserialize the data from the given start_index
+        """
+        # Make sure we have sufficient data for inferring length of list
+        if len(data) < start_index + self.LENGTH_BYTES:
+            raise DeserializationError(
+                'Insufficient data: Cannot retrieve the length of list',
+                data
+            )
+
+        # Number of bytes of only the list data, excluding the prepended list length
+        list_length = int.from_bytes(data[start_index:start_index + self.LENGTH_BYTES], 'big')
+        list_end_index = start_index + self.LENGTH_BYTES + list_length
+        # Make sure we have sufficent data for inferring the whole list
+        if len(data) < list_end_index:
+            raise DeserializationError(
+                'Insufficient data: Cannot retrieve the whole list data',
+                data
+            )
+
+        deserialized_list = []
+        # element_start_index is the start index of an element in the serialized bytes string
+        element_start_index = start_index + self.LENGTH_BYTES
+        while element_start_index < list_end_index:
```

Comment:
This `while` loop makes me *mildly* uncomfortable.  It'd be really nice to figure out a way to have stronger guarantees about the exit conditions.  One that I can think of would be to enforce that each subsequent `element_start_index` **must** be greater than the previous one, so that we have a strong guarantee that this will eventually hit the end of the `data`.